### PR TITLE
Forms

### LINKED ISSUE #23: Bring To/FromFormUrlEncoded from servant
Probably with a generic deriving mechanism for records, whose fields have `To/FromHttpApiData` instances.

### PR DESCRIPTION
**Pull request by fizruk**
Closes #23. Supersedes #31 and #25.

Some further questions:
- do we want `FormOptions` to mimic `aeson`'s `Options`?
- do we want `ToFormKey`/`FromFormKey` instances for `Maybe a`, `Either a b`, `First a`, `Last a` like we have for `ToHttpApiData` and `FromHttpApiData`?
- do we want specialised instances for `Map Text Text`, `HashMap Text Text` and `[(Text, Text)]`?
- do we want to bikeshed functions names (helpers in particular)?

/cc @jkarni @cdepillabout

### CODE REVIEW

File: src/Web/Internal/FormUrlEncoded.hs [hunk 1]
```diff
@@ -0,0 +1,499 @@
...
+instance ToFormKey All where toFormKey = toQueryParam
+instance ToFormKey Any where toFormKey = toQueryParam
+
+instance ToFormKey a => ToFormKey (Dual a)    where toFormKey = toFormKey . getDual
+instance ToFormKey a => ToFormKey (Sum a)     where toFormKey = toFormKey . getSum
+instance ToFormKey a => ToFormKey (Product a) where toFormKey = toFormKey . getProduct
+
+-- | Typeclass for types that can be parsed from keys of a 'Form'. This is the reverse of 'ToFormKey'.
+class FromFormKey k where
+  -- | Parse a key of a 'Form'.
```
Comment (jkarni) [inline src/Web/Internal/FormUrlEncoded.hs]: Looking at the W3 docs, it seems keys (`control names`, in the lingo) should be case-insensitive.

https://www.w3.org/TR/html401/interact/forms.html#control-name
https://www.w3.org/TR/html401/interact/forms.html#h-17.3

So maybe this needs to be `CI Text`?

File: src/Web/Internal/FormUrlEncoded.hs [hunk 2]
```diff
@@ -0,0 +1,499 @@
...
+-- ""
+--
+-- Everything is escaped with @'escapeURIString' 'isUnreserved'@:
+--
+-- >>> encodeForm [("fullname", "Andres Löh")]
+-- "fullname=Andres%20L%C3%B6h"
+encodeForm :: Form -> BSL.ByteString
+encodeForm xs = BSL.intercalate "&" $ map (BSL.fromStrict . encodePair) $ toList xs
+  where
+    escape = Text.encodeUtf8 . Text.pack . escapeURIString isUnreserved . Text.unpack
```
Comment (jkarni) [inline src/Web/Internal/FormUrlEncoded.hs]: This is obviously inefficient. I'm in favor of making changes incrementally, and so leaving the fixing of it for another PR, but for reference [`uri-bytestring`s]((https://hackage.haskell.org/package/uri-bytestring-0.2.2.0/docs/URI-ByteString.html) could ) `urlEncode` could be a good option.

File: src/Web/Internal/FormUrlEncoded.hs [hunk 3]
```diff
@@ -0,0 +1,499 @@
...
+-- (This prevents @'decodeForm' . 'encodeForm'@ from being a true isomorphism).
+--
+-- >>> encodeForm [("", "")]
+-- ""
+--
+-- Everything is escaped with @'escapeURIString' 'isUnreserved'@:
+--
+-- >>> encodeForm [("fullname", "Andres Löh")]
+-- "fullname=Andres%20L%C3%B6h"
+encodeForm :: Form -> BSL.ByteString
```
Comment (jkarni) [inline src/Web/Internal/FormUrlEncoded.hs]: Maybe this should be named `urlEncodeForm`? It's not the only way of encoding forms...

File: src/Web/Internal/FormUrlEncoded.hs [hunk 4]
```diff
@@ -0,0 +1,499 @@
...
+-- Everything is un-escaped with 'unEscapeString':
+--
+-- >>> decodeForm "fullname=Andres%20L%C3%B6h"
+-- Right (fromList [("fullname","Andres L\246h")])
+--
+-- Improperly formed strings result in an error:
+--
+-- >>> decodeForm "this=has=too=many=equals"
+-- Left "not a valid pair: this=has=too=many=equals"
+decodeForm :: BSL.ByteString -> Either Text Form
```
Comment (jkarni) [inline src/Web/Internal/FormUrlEncoded.hs]: Maybe `urlDecodeForm` (see above)?

File: src/Web/Internal/FormUrlEncoded.hs [hunk 5]
```diff
@@ -0,0 +1,499 @@
...
+lookupKey key = concat . HashMap.lookup key . unForm
+
+-- | Lookup a unique value for a key.
+-- Fail if there is zero or more than one value.
+--
+-- >>> lookupUnique "name" []
+-- Left "Could not find key \"name\""
+-- >>> lookupUnique "name" [("name", "Oleg")]
+-- Right "Oleg"
+-- >>> lookupUnique "name" [("name", "Oleg"), ("name", "David")]
```
Comment (jkarni) [inline src/Web/Internal/FormUrlEncoded.hs]: I like the examples!

File: src/Web/Internal/FormUrlEncoded.hs (hunk 2 above)
Comment (fizruk) [inline src/Web/Internal/FormUrlEncoded.hs]: Aha, I was looking for something like it!
[`urlEncodeQuery`](https://hackage.haskell.org/package/uri-bytestring-0.2.2.0/docs/URI-ByteString.html#v:urlEncodeQuery) looks ideal for encoding.

File: src/Web/Internal/FormUrlEncoded.hs (hunk 3 above)
Comment (fizruk) [inline src/Web/Internal/FormUrlEncoded.hs]: Yes, `urlEncodeForm` sounds ok.

**Comment by jkarni** [comment]
What did you decide about sum types? (Looking through the code, I remember that I settled for having them when they're.)

**Comment by fizruk** [comment]
@jkarni ah, I still want to remove support for them for now. If there's demand we'll get back to this problem later.

File: src/Web/Internal/FormUrlEncoded.hs (hunk 1 above)
Comment (fizruk) [inline src/Web/Internal/FormUrlEncoded.hs]: Interesting, we might want to use `CI Text` for helpers in `Web.HttpApiData` too.

Comment (fizruk) [inline src/Web/Internal/FormUrlEncoded.hs]: @jkarni can't find anywhere that "control names" should be case insensitive.

Comment (jkarni) [inline src/Web/Internal/FormUrlEncoded.hs]: I was looking at `name` in [17.3](https://www.w3.org/TR/html401/interact/forms.html#h-17.3), but that's not right. `id` says it's case sensitive. So I think it's right as it stands.

**Comment by cdepillabout** [comment]
@fizruk: This looks good to me.

One thing I'd like to see is an example using `FormOptions`.  One thing that always bugs me is when I look at lens or aeson documentation is that there is not enough explanation for using their generic deriving options.

For example, take a datatype like this:

```
data Person = Person
  { personName :: Text
  , personAge :: Int
  }
```

I imagine it's common to want to urlencode this to be `name=foo&age=30`, not `personName=foo&personAge=30`.

An example of how to use `defaultOptions` to accomplish this would be really nice.

Of course, I can always send a PR for this additional documentation after this PR (#32) has been merged.

**Comment by fizruk** [comment]
@cdepillabout added an example in https://github.com/fizruk/http-api-data/pull/32/commits/e613d050e206add3bca1b7acaf98e997b444ff4a.

Any final wishes before I release? :)

**Comment by jkarni** [comment]
I wish you a good flight?

**Comment by fizruk** [comment]
@jkarni yeah, thanks :)

I am merging this now, I think we've covered most things here.

**Pull request merged**